Acesso ao Sistema Expectativas

angelosalton opened the pull request:
Este pull request implementa uma função para acesso aos dados do [Sistema de Expectativas de Mercado](https://www3.bcb.gov.br/expectativas/publico/consulta/serieestatisticas) do Banco Central.
- Optei por criar um script separado para não causar confusão com as funcionalidades atuais do pacote.
- Os testes unitários ainda não cobrem todos os casos possíveis de uso da função.

Commit:
Added call to BCB Olinda API

Commit:
Added expectations to test suite

Commit:
Initial interface to expectations

opardal reviewed (changes requested):
Refiz a revisão, conforme post abaixo.

opardal commented on sgs/api.py:
```diff
@@ -22,3 +23,35 @@ def get_data(ts_code: int, begin: str, end: str) -> List:
+        }
+
+    # TODO: Olinda API expects ISO dates
+    begin_iso = datetime.datetime.strptime(begin, '%d/%m/%Y').strftime('%Y-%m-%d')
```
```suggestion
    begin_iso = to_datetime(begin, 'pt')
```

opardal commented on sgs/api.py:
```diff
@@ -1,3 +1,4 @@
+import datetime
```
Não vejo necessidade de importar o módulo datetime. A atual versão deste arquivo já faz referência a função to_datetime do módulo common.py.

```suggestion
```

Sugiro alterar a linha 8 para:
```
from .common import LRU_CACHE_SIZE, MAX_ATTEMPT_NUMBER, to_datetime
``` 
Dessa forma, reaproveitamos a função `to_datetime`.

opardal commented on sgs/api.py:
```diff
@@ -22,3 +23,35 @@ def get_data(ts_code: int, begin: str, end: str) -> List:
+        }
+
+    # TODO: Olinda API expects ISO dates
+    begin_iso = datetime.datetime.strptime(begin, '%d/%m/%Y').strftime('%Y-%m-%d')
+    end_iso = datetime.datetime.strptime(end, '%d/%m/%Y').strftime('%Y-%m-%d')
```
Sugiro alterar o seguinte trecho para referenciar a função já importada:

```suggestion
    begin_iso = to_datetime(begin, 'pt')
    end_iso = to_datetime(end, 'pt')
```

opardal commented on sgs/expectations.py:
```diff
@@ -0,0 +1,31 @@
+from . import api
+
+import numpy as np
```
Módulo não utilizado.

```suggestion

```

opardal commented on tests/test_expectations.py:
```diff
@@ -0,0 +1,20 @@
+import numpy as np
```
Módulo não utilizado.

```suggestion
```

opardal commented on tests/test_expectations.py:
```diff
@@ -0,0 +1,20 @@
+
+
+@pytest.mark.expectations
+def test_expectations_invalid_resource():
+    actual = expectations("foobar", "01/01/2020", "01/02/2020")
+    expected = None
+
+    assert actual == expected
```
Alterar o teste test_expectations_invalid_resource. A função, quando alimentada com um resource inválido, retorna um ValueError

```suggestion
def test_expectations_invalid_resource():
    with pytest.raises(ValueError):
        expectations("foobar", "01/01/2020", "01/02/2020")
```

opardal reviewed (changes requested):
Ola, @angelosalton 

Tenho algumas sugestões para seu PR:

**1) Arquivo `__init__.py`**

Incluir o seguinte trecho para fazermos a referência:
```
from .expectations import expectations
```

**2) Arquivo `api.py`**

Não vejo necessidade de importar o módulo `datetime`. A atual versão deste arquivo já faz referência a função `to_datetime` do módulo `common.py`.

Portanto, incluir na linha de importação da `common.py` a função `to_datetime`. Resultando na seguinte linha de código:
```
from .common import LRU_CACHE_SIZE, MAX_ATTEMPT_NUMBER, to_datetime
``` 

Assim, sugiro alterar o seguinte trecho para referenciar a função já importada:

De:
```
 begin_iso = datetime.datetime.strptime(begin, '%d/%m/%Y').strftime('%Y-%m-%d')
 end_iso = datetime.datetime.strptime(end, '%d/%m/%Y').strftime('%Y-%m-%d')
```
Para:
```
 begin_iso = to_datetime(begin, 'pt')
 end_iso = to_datetime(end, 'pt')
```  

**3) Arquivo `expectations.py`**

Remover a importação do modulo `numpy`.


**4) Arquivo `test_expectations.py`**

Remover a importação do modulo `numpy`.

Alterar o teste `test_expectations_invalid_resource`. A função, quando alimentada com um `resource` inválido, retorna um `ValueError`. Assim, sugiro alterar o teste para:
```
@pytest.mark.expectations
def test_expectations_invalid_resource():
    with pytest.raises(ValueError):
        expectations("foobar", "01/01/2020", "01/02/2020")
``` 
Qualquer dúvida, estou a disposição para qualquer tipo de auxílio. Abs.

Commit:
Adicionar sugestões de code review

@opardal, obrigado pelas suas sugestões!
É meu primeiro PR aqui no pacote, futuramente revisarei os códigos com mais calma.

Co-authored-by: O Pardal <js.cad@pm.me>

Commit:
Merge branch 'develop' into expectativas

opardal commented:
Oi, @angelosalton 

O seu PR não está passando nos checks ainda.

Veja, por favor, se modificando o arquivo  **`__init__.py`** dentro do diretorio sgs resolve.

Sugestão: incluir o seguinte trecho:
```
from .expectations import expectations
```

opardal committed:
Expectativas (#2)

* Update pytest.ini

* Update __init__.py

* Update api.py

Manipulando o caso em que a `end = None`

* Update api.py

Corrigido um erro no uso da função `to_datetime`

* Update api.py

Removida a possibilidade de utilizar `end = None`

* Update api.py

Retomando a abordagem com a biblioteca datetime.

A função to_datetime gera conflitos.

* Update api.py

opardal commented on pytest.ini:
```diff
@@ -10,3 +10,4 @@ markers =
     mypy
     search
     ts
```
```suggestion
    ts strict
```

opardal replied:
```suggestion
    ts
    strict
```

opardal commented:
@rafpyprog : Quando vamos incluir esse trabalho aqui? Acho que o @angelosalton vai ter que dar um rebase ou nao precisa?
